langchain/src/vectorstores/typeorm.ts

```diff
@@ -0,0 +1,307 @@
+        .map((d) => d.id);
+
+      this.verboseConsoleLog(
+        `Ids has changed for document ${doc.sourceName} and ${doc.sourceType}, action: delete this ids: ${idsDocumentsThatWeShouldDeleted}`
+      );
+
+      await documentRepository.delete({
+        id: {
+          in: idsDocumentsThatWeShouldDeleted,
+        },
+      });
+    }
+
+    const documentsToUpsert = [];
+    for (const row of rows) {
+      // Search for existing document with same sourceName and sourcetype
+      const documentsInDatabase = (await documentRepository.findOne({
+        where: {
+          id: row.id,
+        },
+      })) as Document;
+
+      if (documentsInDatabase) {
+        if (documentsInDatabase.hash === row.hash) {
+          this.verboseConsoleLog(
+            `Same hash was found for id ${row.id}, action: ignore`
+          );
+        } else {
+          documentsToUpsert.push(row);
+          this.verboseConsoleLog(
+            `Different hash was found for id ${row.id}, action: update embedding`
+          );
+        }
+      } else {
+        documentsToUpsert.push(row);
+        this.verboseConsoleLog(
+          `No document was found for id ${row.id}, action: create`
+        );
+      }
+    }
+
+    const chunkSize = 500;
+    for (let i = 0; i < documentsToUpsert.length; i += chunkSize) {
+      const chunk = documentsToUpsert.slice(i, i + chunkSize);
+
+      try {
+        await documentRepository.upsert(chunk, [`id`]);
+      } catch (e) {
+        console.error(e);
+        throw new Error(`Error inserting: ${chunk[0].pageContent}`);
+      }
+    }
+  }
+
+  async similaritySearchVectorWithScore(
+    query: number[],
+    k: number,
+    filter?: this["FilterType"]
+  ): Promise<[TypeORMVectorStoreDocument, number][]> {
+    const embeddingString = `'[${query.join(",")}]'`;
+    const _filter = filter ?? "'{}'";
+    const queryString = `
+      SELECT *, embedding <=> ${embeddingString}::vector as "_distance" 
+      FROM ${this.tableName}
+      WHERE metadata @> ${_filter}
+      ORDER BY "_distance" ASC
+      LIMIT ${k};
+    `;
+    const documents = await this.appDataSource.query(queryString);
+
+    const results = [] as [TypeORMVectorStoreDocument, number][];
+    for (const doc of documents) {
+      if (doc._distance != null && doc.pageContent != null) {
+        const document = new Document(doc) as TypeORMVectorStoreDocument;
+        document.id = doc.id;
+        results.push([document, doc._distance]);
+      }
+    }
+
+    return results;
+  }
+
+  async initTable(): Promise<void> {
+    await this.appDataSource.query("CREATE EXTENSION IF NOT EXISTS vector;");
+
+    await this.appDataSource.query(`
+      CREATE TABLE IF NOT EXISTS ${this.tableName} (
+        id text PRIMARY KEY,
+        "pageContent" text,
+        metadata jsonb,
+        "sourceType" text,
+        "sourceName" text,
+        hash text,
+        embedding vector
+      );
+    `);
+  }
+
+  static async fromTexts(
```

Comment:
I add a split of the texts because either the user has already done it and it should be transparent or it hasn't and it could lead to errors. I had people asking for my help because of this.

Reply:
This would be a broader problem right? Let's make this a separate PR and try to think of a more general solution. There are a lot of different splitting strategies.

Reply:
OK I removed it